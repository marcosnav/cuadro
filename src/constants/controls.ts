/**
 * Available control types
 */
export enum ControlType {
  NEW_GAME = 'NEW_GAME',
  RESTART_GAME = 'RESTART_GAME',
  SEE_ORIGINAL = 'SEE_ORIGINAL',
}
