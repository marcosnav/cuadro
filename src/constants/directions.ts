/**
 * SwipeDirection, one of UP, RIGHT, LEFT, DOWN
 */
export enum SwipeDirection {
  UP = 'UP',
  RIGHT = 'RIGHT',
  DOWN = 'DOWN',
  LEFT = 'LEFT',
}
