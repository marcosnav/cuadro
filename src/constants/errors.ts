export enum errors {
  NO_INTERNET = 'NO_INTERNET',
  UNEXPECTED_ERROR = 'UNEXPECTED_ERROR',
  UNRESPONSIVE_IMAGE_PROVIDER = 'UNRESPONSIVE_IMAGE_PROVIDER',
}
