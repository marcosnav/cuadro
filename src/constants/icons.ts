import check from './../assets/svg/check.svg';
import image from './../assets/svg/image.svg';
import newicon from './../assets/svg/new.svg';
import refresh from './../assets/svg/refresh.svg';

export const icons: { [key: string]: string } = {
  check,
  image,
  new: newicon,
  refresh,
};
