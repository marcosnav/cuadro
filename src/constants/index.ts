export * from './status';
export * from './errors';
export * from './theme';
export * from './keycodes';
