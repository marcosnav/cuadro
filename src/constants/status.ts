export enum status {
  LOADING_IMAGE = 'LOADING_IMAGE',
  DISPLAY_ORIGINAL = 'DISPLAY_ORIGINAL',
  PLAYING_GAME  = 'PLAYING_GAME ',
  FINISHED_GAME = 'FINISHED_GAME',
}
