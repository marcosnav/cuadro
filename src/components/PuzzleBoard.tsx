import React, { FC, useCallback, useEffect, useState } from 'react';
import { SwipeDirection } from './../constants';
import PuzzlePiece from './PuzzlePiece';
import * as S from './styled';

interface IProps {
  image: string;
  puzzleState: number[];
  onSwipe: (direction: SwipeDirection) => void;
}

const swipeThreshold = 50;

/**
 * PuzzleBoard Component
 * {puzzleState} The current matrix of the board pieces
 * {onSwipe} Function Handler that recieves SwipeDirection as parameter
 * @param props {puzzleState: number[][], onSwipe: Function}
 */
const PuzzleBoard: FC<IProps> = ({ image, puzzleState, onSwipe }) => {
  const [startX, setStartX] = useState(0);
  const [startY, setStartY] = useState(0);
  const [endX, setEndX] = useState(0);
  const [endY, setEndY] = useState(0);

  const handleSwipe = useCallback(() => {
    const xDistance = endX - startX;
    const yDistance = endY - startY;
    const absX = Math.abs(xDistance);
    const absY = Math.abs(yDistance);
    const { UP, DOWN, LEFT, RIGHT } = SwipeDirection;
    let direction: SwipeDirection|null = null;

    if (absX < swipeThreshold && absY < swipeThreshold) {
      return;
    }

    if (absX > absY) {
      direction = xDistance > 0 ? RIGHT : LEFT;
    } else {
      direction = yDistance > 0 ? DOWN :  UP;
    }

    if (direction) {
      onSwipe(direction);
    }
  }, [endX, endY, onSwipe]);

  const startSwipe = (ev: React.TouchEvent) => {
    ev.stopPropagation();
    setStartX(ev.touches[0].screenX);
    setStartY(ev.touches[0].screenY);
  };

  const endSwipe = (ev: React.TouchEvent) => {
    ev.stopPropagation();
    setEndX(ev.changedTouches[0].screenX);
    setEndY(ev.changedTouches[0].screenY);
  };

  useEffect(() => {
    handleSwipe();
  }, [handleSwipe]);

  const boardPieces = () => {
    return puzzleState.map((pos: number, index) => {
      if (!index) {
        return null;
      }
      return <PuzzlePiece key={index} image={image} num={index} position={pos} />;
    });
  };

  return (
    <S.Board onTouchStart={startSwipe} onTouchEnd={endSwipe}>
      <S.PuzzleHolder>
        {boardPieces()}
      </S.PuzzleHolder>
    </S.Board>
  );
};

export default PuzzleBoard;
