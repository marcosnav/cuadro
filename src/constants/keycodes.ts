import { SwipeDirection } from './directions';

/**
 * KeyCodes to match for keypress events
 */
export enum KeyCodes {
  DOWN = 40,
  LEFT = 37,
  RIGHT = 39,
  UP = 38,
  ENTER = 13,
}

/**
 * Map key code with swipe direction
 */
export const KeySwipeMap: { [key: number]: SwipeDirection } = {
  40: SwipeDirection.DOWN,
  37: SwipeDirection.LEFT,
  39: SwipeDirection.RIGHT,
  38: SwipeDirection.UP,
};
