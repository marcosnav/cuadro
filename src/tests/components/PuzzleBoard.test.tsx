import { mount } from 'enzyme';
import React from 'react';
import PuzzleBoard from './../../components/PuzzleBoard';
import { SwipeDirection } from './../../constants';

const puzzleStateMock = [1, 0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];

const touchHandler = {
  move: (x: number, y: number) => ({ changedTouches: [{ screenX: x, screenY: y }] }),
  start: (x: number, y: number) => ({ touches: [{ screenX: x, screenY: y }] }),
};

describe('<PuzzleBoard/>', () => {
  test('renders properly', () => {
    const swipedEvMock = () => '';
    const component = mount(
    <PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedEvMock} showOriginal={false} />);
    expect(component).toMatchSnapshot();
  });

  describe('onSwipe', () => {
    const swipedSpy = jest.fn();
    const component = mount(
    <PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} showOriginal={false} />);

    test('not enough swipe distance', () => {
      component.simulate('touchStart', touchHandler.start(150, 150));
      component.simulate('touchEnd', touchHandler.move(110, 180));
      expect(swipedSpy).not.toHaveBeenCalled();
    });

    test('swipe left, calls with SwipeDirection.LEFT', () => {
      component.simulate('touchStart', touchHandler.start(150, 150));
      component.simulate('touchEnd', touchHandler.move(50, 150));
      expect(swipedSpy).toHaveBeenCalledTimes(1);
      expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.LEFT);
    });

    test('swipe right, calls with SwipeDirection.RIGHT', () => {
      component.simulate('touchStart', touchHandler.start(150, 150));
      component.simulate('touchEnd', touchHandler.move(250, 150));
      expect(swipedSpy).toHaveBeenCalledTimes(2);
      expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.RIGHT);
    });

    test('swipe up, calls with SwipeDirection.UP', () => {
      component.simulate('touchStart', touchHandler.start(150, 150));
      component.simulate('touchEnd', touchHandler.move(150, 50));
      expect(swipedSpy).toHaveBeenCalledTimes(3);
      expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.UP);
    });

    test('swipe down, calls with SwipeDirection.DOWN', () => {
      component.simulate('touchStart', touchHandler.start(150, 150));
      component.simulate('touchEnd', touchHandler.move(150, 250));
      expect(swipedSpy).toHaveBeenCalledTimes(4);
      expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.DOWN);
    });
  });

  describe('on ambiguous diagonal swiping', () => {
    describe('longest distance must define swiped side', () => {
      test('left > up', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(50, 60));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.LEFT);
      });

      test('left < up', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(60, 50));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.UP);
      });

      test('right > up', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(260, 60));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.RIGHT);
      });

      test('right < up', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(240, 50));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.UP);
      });

      test('left > down', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(50, 240));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.LEFT);
      });

      test('left < down', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(60, 260));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.DOWN);
      });

      test('right > down', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(260, 240));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.RIGHT);
      });

      test('right < down', () => {
        const swipedSpy = jest.fn();
        const component = mount(<PuzzleBoard image={''} puzzleState={puzzleStateMock} onSwipe={swipedSpy} />);
        component.simulate('touchStart', touchHandler.start(150, 150));
        component.simulate('touchEnd', touchHandler.move(240, 260));
        expect(swipedSpy).toHaveBeenCalledTimes(1);
        expect(swipedSpy).toHaveBeenCalledWith(SwipeDirection.DOWN);
      });
    });
  });
});
