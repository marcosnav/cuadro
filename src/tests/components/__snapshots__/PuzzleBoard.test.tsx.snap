// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PuzzleBoard/> renders properly 1`] = `
<Board
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="sc-iwsKbI epIAEG"
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <PuzzleHolder>
      <div
        className="sc-gZMcBi gDgmOB"
      >
        <Piece
          left={0}
          top={0}
        >
          <div
            className="sc-gqjmRU jHsaGe"
          >
            <PieceImage
              image=""
              left={-67}
              top={-0}
            >
              <div
                className="sc-VigVT fIqCJj"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={134}
          top={0}
        >
          <div
            className="sc-gqjmRU guMLcV"
          >
            <PieceImage
              image=""
              left={-134}
              top={-0}
            >
              <div
                className="sc-VigVT eFTdJG"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={201}
          top={0}
        >
          <div
            className="sc-gqjmRU dukDdZ"
          >
            <PieceImage
              image=""
              left={-201}
              top={-0}
            >
              <div
                className="sc-VigVT fFWRCF"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={0}
          top={67}
        >
          <div
            className="sc-gqjmRU dzpEya"
          >
            <PieceImage
              image=""
              left={-0}
              top={-67}
            >
              <div
                className="sc-VigVT jljySD"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={67}
          top={67}
        >
          <div
            className="sc-gqjmRU kbLygx"
          >
            <PieceImage
              image=""
              left={-67}
              top={-67}
            >
              <div
                className="sc-VigVT ctaygX"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={134}
          top={67}
        >
          <div
            className="sc-gqjmRU bfgdvl"
          >
            <PieceImage
              image=""
              left={-134}
              top={-67}
            >
              <div
                className="sc-VigVT fgDWPo"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={201}
          top={67}
        >
          <div
            className="sc-gqjmRU hXXxcB"
          >
            <PieceImage
              image=""
              left={-201}
              top={-67}
            >
              <div
                className="sc-VigVT zikom"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={0}
          top={134}
        >
          <div
            className="sc-gqjmRU hHDSto"
          >
            <PieceImage
              image=""
              left={-0}
              top={-134}
            >
              <div
                className="sc-VigVT htULSm"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={67}
          top={134}
        >
          <div
            className="sc-gqjmRU bzclvp"
          >
            <PieceImage
              image=""
              left={-67}
              top={-134}
            >
              <div
                className="sc-VigVT hvJYcH"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={134}
          top={134}
        >
          <div
            className="sc-gqjmRU fVeBUQ"
          >
            <PieceImage
              image=""
              left={-134}
              top={-134}
            >
              <div
                className="sc-VigVT lmZoYJ"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={201}
          top={134}
        >
          <div
            className="sc-gqjmRU dKmhpK"
          >
            <PieceImage
              image=""
              left={-201}
              top={-134}
            >
              <div
                className="sc-VigVT kWGaLb"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={0}
          top={201}
        >
          <div
            className="sc-gqjmRU jcWuJk"
          >
            <PieceImage
              image=""
              left={-0}
              top={-201}
            >
              <div
                className="sc-VigVT EPtpx"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={67}
          top={201}
        >
          <div
            className="sc-gqjmRU bduzEv"
          >
            <PieceImage
              image=""
              left={-67}
              top={-201}
            >
              <div
                className="sc-VigVT kYveBR"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={134}
          top={201}
        >
          <div
            className="sc-gqjmRU jKBdKP"
          >
            <PieceImage
              image=""
              left={-134}
              top={-201}
            >
              <div
                className="sc-VigVT ixgRcQ"
              />
            </PieceImage>
          </div>
        </Piece>
        <Piece
          left={201}
          top={201}
        >
          <div
            className="sc-gqjmRU hYfMog"
          >
            <PieceImage
              image=""
              left={-201}
              top={-201}
            >
              <div
                className="sc-VigVT ktZbvi"
              />
            </PieceImage>
          </div>
        </Piece>
      </div>
    </PuzzleHolder>
  </div>
</Board>
`;
